README.md
# pretty-units
[![Build Status][travis-image]][travis-url]
[![NPM][npm-image]][npm-url]

Have you received a measurement that had an unsightly or unpredictable order of magnitude?

*pretty-units* will find a number's largest satisfiable order of magnitude and provide a string representation that includes both the number and a unit prefix. All you need to do is append your unit!

*pretty-units* supports transformation from *yocto* (10<sup>-24</sup>) to *yotta* (10<sup>24</sup>) and everything in between!

Output is always rounded to two decimal places. Naturally, this means that *pretty-units* specially handles values that have an absolute value that is less than 10<sup>-26</sup>. See [usage](#usage) for more details.

## <a name="usage"></a>Usage

```js
> var punits = require('pretty-units');

> punits(9000) + 'm';
'9 km'

> punits(0.001) + 'l';
'1 ml'

> punits(Math.pow(10, 12)) + 's';
'1 Ts'

// use a custom exponent of 10
> punits(-0.9, 3) + 'm';
'-900 m'

> punits(2, -3) + 'l';
'2 ml'

> punits(9.235, 9) + 'B';
'9.24 GB'

// let's handle some very small numbers
> punits(1, -26) + 's'
'0.01 ys'

> punits(-1, -27) + 's'
'< -0.01 ys'

> punits(0.99 * Math.pow(10, -26)) + 's'
'0.01 ys'

> punits(0.4 * Math.pow(10, -26)) + 's'
'< 0.01 ys'

// switch to verbose output
> punits.verbose = true;

> punits(9000) + 'meters';
'9 kilometers'

> punits(0.001) + 'liter';
'1 milliliter'

```

[travis-image]: https://img.shields.io/travis/davidgwking/pretty-units.svg?style=flat&branch=master
[travis-url]: https://travis-ci.org/davidgwking/pretty-units
[npm-image]: http://img.shields.io/npm/v/pretty-units.svg
[npm-url]: https://www.npmjs.com/package/pretty-units


CHANGELOG.md
0.1.0 / 2015-02-07
==================
* iterator performance enhancement
* refactored some behavior from `prett-units.js` to `prefixes.js`

0.0.2 / 2015-02-06
==================
* powers of ten are now calculated once

0.0.1 / 2015-02-06
==================
* initial release


test/pretty-units.js
var expect = require('chai').expect;
var punits = require('../lib/pretty-units');

describe('pretty-units', function () {

  it('should format values', function () {
    expect(punits(Math.pow(10, -26))).to.eql('0.01 y');
    expect(punits(Math.pow(10, -25))).to.eql('0.1 y');
    expect(punits(Math.pow(10, -24))).to.eql('1 y');
    expect(punits(Math.pow(10, -23))).to.eql('10 y');
    expect(punits(Math.pow(10, -22))).to.eql('100 y');
    expect(punits(Math.pow(10, -21))).to.eql('1 z');
    expect(punits(Math.pow(10, -20))).to.eql('10 z');
    expect(punits(Math.pow(10, -19))).to.eql('100 z');
    expect(punits(Math.pow(10, -18))).to.eql('1 a');
    expect(punits(Math.pow(10, -17))).to.eql('10 a');
    expect(punits(Math.pow(10, -16))).to.eql('100 a');
    expect(punits(Math.pow(10, -15))).to.eql('1 f');
    expect(punits(Math.pow(10, -14))).to.eql('10 f');
    expect(punits(Math.pow(10, -13))).to.eql('100 f');
    expect(punits(Math.pow(10, -12))).to.eql('1 p');
    expect(punits(Math.pow(10, -11))).to.eql('10 p');
    expect(punits(Math.pow(10, -10))).to.eql('100 p');
    expect(punits(Math.pow(10, -9))).to.eql('1 n');
    expect(punits(Math.pow(10, -8))).to.eql('10 n');
    expect(punits(Math.pow(10, -7))).to.eql('100 n');
    expect(punits(Math.pow(10, -6))).to.eql('1 μ');
    expect(punits(Math.pow(10, -5))).to.eql('10 μ');
    expect(punits(Math.pow(10, -4))).to.eql('100 μ');
    expect(punits(Math.pow(10, -3))).to.eql('1 m');
    expect(punits(Math.pow(10, -2))).to.eql('10 m');
    expect(punits(Math.pow(10, -1))).to.eql('100 m');
    expect(punits(Math.pow(10, 0))).to.eql('1 ');
    expect(punits(Math.pow(10, 1))).to.eql('10 ');
    expect(punits(Math.pow(10, 2))).to.eql('100 ');
    expect(punits(Math.pow(10, 3))).to.eql('1 k');
    expect(punits(Math.pow(10, 4))).to.eql('10 k');
    expect(punits(Math.pow(10, 5))).to.eql('100 k');
    expect(punits(Math.pow(10, 6))).to.eql('1 M');
    expect(punits(Math.pow(10, 7))).to.eql('10 M');
    expect(punits(Math.pow(10, 8))).to.eql('100 M');
    expect(punits(Math.pow(10, 9))).to.eql('1 G');
    expect(punits(Math.pow(10, 10))).to.eql('10 G');
    expect(punits(Math.pow(10, 11))).to.eql('100 G');
    expect(punits(Math.pow(10, 12))).to.eql('1 T');
    expect(punits(Math.pow(10, 13))).to.eql('10 T');
    expect(punits(Math.pow(10, 14))).to.eql('100 T');
    expect(punits(Math.pow(10, 15))).to.eql('1 P');
    expect(punits(Math.pow(10, 16))).to.eql('10 P');
    expect(punits(Math.pow(10, 17))).to.eql('100 P');
    expect(punits(Math.pow(10, 18))).to.eql('1 E');
    expect(punits(Math.pow(10, 19))).to.eql('10 E');
    expect(punits(Math.pow(10, 20))).to.eql('100 E');
    expect(punits(Math.pow(10, 21))).to.eql('1 Z');
    expect(punits(Math.pow(10, 22))).to.eql('10 Z');
    expect(punits(Math.pow(10, 23))).to.eql('100 Z');
    expect(punits(Math.pow(10, 24))).to.eql('1 Y');
    expect(punits(Math.pow(10, 25))).to.eql('10 Y');
    expect(punits(Math.pow(10, 26))).to.eql('100 Y');
    expect(punits(Math.pow(10, 27))).to.eql('1000 Y');
    expect(punits(Math.pow(10, 28))).to.eql('10000 Y');
    expect(punits(Math.pow(10, 29))).to.eql('100000 Y');
    expect(punits(Math.pow(10, 30))).to.eql('1000000 Y');
  });

  it('should format values with a custom exponent', function () {
    expect(punits(10, 3)).to.eql('10 k');        // actual: 10000
    expect(punits(1, 4)).to.eql('10 k');         // actual: 10000
    expect(punits(10.23, 3)).to.eql('10.23 k');  // actual: 10230
    expect(punits(12.3, 6)).to.eql('12.3 M');    // actual: 12,300,000
    expect(punits(12.3, -3)).to.eql('12.3 m');   // actual: 0.0123
    expect(punits(12.3, -4)).to.eql('1.23 m');   // actual: 0.00123
  });

  it('should support zero', function () {
    expect(punits(0)).to.eql('0 ');
  });

  it('should support verbose prefixes', function () {
    punits.verbose = true;
    expect(punits(-10, 3)).to.eql('-10 kilo');
    expect(punits(-Math.pow(10, 20), 0, true)).to.eql('-100 exa');
    expect(punits(-12.3, -4, true)).to.eql('-1.23 milli');
    punits.verbose = false;
  });

  it('should round output to two decimal places', function () {
    expect(punits(1006)).to.eql('1.01 k');
    expect(punits(1215.7231)).to.eql('1.22 k');
    expect(punits(10.006, 3)).to.eql('10.01 k');
    expect(punits(Math.pow(10, -27))).to.eql('< 0.01 y');
  });

  it('should support negative numbers', function () {
    expect(punits(-10, 3)).to.eql('-10 k');
    expect(punits(-Math.pow(10, 20))).to.eql('-100 E');
    expect(punits(-12.3, -4)).to.eql('-1.23 m');
    expect(punits(-Math.pow(10, -27))).to.eql('< -0.01 y');
  });

});


lib/pretty-units.js
var prefixes = require('./prefixes');

/**
 * Pretty format a number with SI prefixes.
 *
 * @param {Number}  number        a number
 * @param {Number}  exp           custom exponent of 10
 */
module.exports = function prettyUnits(number, exp) {
  exp = exp || 0;
  number = exp ? number * Math.pow(10, exp) : number;

  // immediately return for zero
  if (number === 0) return '0 ';

  // adjust number
  var prefixIndex = prefixes.getPrefixIndex(number);
  var adjusted = prefixes.adjustNumber(number, prefixIndex);

  // provide some failover for small numbers (not support due to 2 decimal rounding scheme)
  if (adjusted === 0) adjusted = '< ' + (number > 0 ? '' : '-') + '0.01';

  // get my prefix
  var prefix = prefixes[prefixIndex][prefixProp];

  return adjusted + ' ' + prefix;
};

var verbose = false;
var prefixProp = 'symbol';
Object.defineProperty(module.exports, 'verbose', {
  set: function(bool) {
    verbose = !!bool;
    prefixProp = verbose ? 'prefix' : 'symbol';
  }
});


lib/prefixes.js
var util = require('./util');

/**
 * SI Prefix Dictionary
 */
var prefixes =  [
  {pow10: Math.pow(10, -24),  prefix: 'yocto',  symbol: 'y'},
  {pow10: Math.pow(10, -21),  prefix: 'zepto',  symbol: 'z'},
  {pow10: Math.pow(10, -18),  prefix: 'atto',   symbol: 'a'},
  {pow10: Math.pow(10, -15),  prefix: 'femto',  symbol: 'f'},
  {pow10: Math.pow(10, -12),  prefix: 'pico',   symbol: 'p'},
  {pow10: Math.pow(10, -9),   prefix: 'nano',   symbol: 'n'},
  {pow10: Math.pow(10, -6),   prefix: 'micro',  symbol: 'μ'},
  {pow10: Math.pow(10, -3),   prefix: 'milli',  symbol: 'm'},
  {pow10: Math.pow(10, 0),    prefix: '',       symbol: ''},
  {pow10: Math.pow(10, 3),    prefix: 'kilo',   symbol: 'k'},
  {pow10: Math.pow(10, 6),    prefix: 'mega',   symbol: 'M'},
  {pow10: Math.pow(10, 9),    prefix: 'giga',   symbol: 'G'},
  {pow10: Math.pow(10, 12),   prefix: 'tera',   symbol: 'T'},
  {pow10: Math.pow(10, 15),   prefix: 'peta',   symbol: 'P'},
  {pow10: Math.pow(10, 18),   prefix: 'exa',    symbol: 'E'},
  {pow10: Math.pow(10, 21),   prefix: 'zetta',  symbol: 'Z'},
  {pow10: Math.pow(10, 24),   prefix: 'yotta',  symbol: 'Y'}
];
module.exports = prefixes;

/**
 * Adjust the magnitude of a number to that indicated by the prefix at prefixIndex.
 */
function adjustNumber(number, prefixIndex) {
  return util.round(number / prefixes[prefixIndex].pow10, 2);
}
module.exports.adjustNumber = adjustNumber;

/**
 * Get the most "best match" prefix index for a given number/
 */
function getPrefixIndex(number) {
  // use absolute value for comparisons
  var abs = Math.abs(number);

  // attempt to match extremities
  if (abs <= prefixes[0].pow10) return 0;
  else if (abs >= prefixes[prefixes.length - 1].pow10) return prefixes.length - 1;

  // find best match
  for (var i = getInitialPrefixIterPos(); i < prefixes.length; i++) {
    if (abs < prefixes[i].pow10) return i - 1;
  }

  return 8;
}
module.exports.getPrefixIndex = getPrefixIndex;

/**
 * A crude optimization so that we can get a good iterator start position.
 *
 * Assumes absNumb is an the absolute value of a number.
 */
function getInitialPrefixIterPos(absNum) {
  return absNum >= prefixes[8].pow10 ? 8 : 0;
}
module.exports.getInitialPrefixIterPos = getInitialPrefixIterPos;
